pyproject.toml

```diff
@@ -194,6 +192,32 @@ python_files = "test_*.py"
 
 norecursedirs = "__pycache__ build .pytest_cache *.egg-info .venv .github-venv"
 
+[tool.tox]
+legacy_tox_ini = """
+    [tox]
+    envlist = py310
+
+    [testenv]
+    # these options make Tox use the existing .venv dir created by venv
+    envdir = {env:VENV:.venv}
+    usedevelop = true
+    recreate = false
+    skip_install = true
+
+    setenv =
+        # configure --cov here because we can't use
+        # envvars in pyproject.toml, and we want this
+        # to be configurable in order to test coverage
+        # for a subset of tests in the monorepo rather
+        # than for all tests in the repo.
+        PYTEST_ADDOPTS = --cov={env:PYTEST_TARGET:.}
+        # additional pytest flags - defaults to no flags
+        PYTEST_FLAGS = {env:PYTEST_FLAGS:}
+    commands =
+        # use `.` add the default --cov option if it's not specified
+        pytest {env:PYTEST_FLAGS} {env:PYTEST_ADDOPTS:--cov=.} {env:PYTEST_TARGET}
```

Comment:
**question (blocking):** Your command is using `PYTEST_FLAGS`, `PYTEST_ADDOPTS`, and `PYTEST_TARGET`, but you're only establishing `PYTEST_FLAGS` and `PYTEST_ADDOPTS` with `setenv`. Is `PYTEST_TARGET` actually getting through?

If not, you can use [`passenv`](https://tox.wiki/en/latest/config.html#pass_env) to list envvars that _should_ pass through tox (you could probably move `PYTEST_FLAGS` there as well).

Reply:
I don't totally remember why I needed to do it this way, but you can also supply those on the command line if you don't want them in `pyproject.toml`:

https://github.com/uclahs-cds/tool-BoutrosLab-utils/blob/98c270d92007f07de32869770b72ad099aeae5c9/.github/workflows/pytest.yaml#L56-L58

```console
          tox \
            -x 'testenv.pass_env=HOME,GITHUB_TOKEN,GIT_ASKPASS' \
            -e py${{matrix.python-version}}
```

Reply:
> **question (blocking):** Your command is using `PYTEST_FLAGS`, `PYTEST_ADDOPTS`, and `PYTEST_TARGET`, but you're only establishing `PYTEST_FLAGS` and `PYTEST_ADDOPTS` with `setenv`. Is `PYTEST_TARGET` actually getting through?
> 

It is - as far as I know, default behavior of `sh` (called through `make`) is to call `vfork` to execute a command, which just clones the existing process, including its env. In testing, `PYTEST_TARGET` is passed through as expected. I defined `PYTEST_ADDOPTS` here because I need the default `--cov=.`. On the other hand, `pytest` doesn't need a default target specified because it defaults to `.` anyway. I am not sure why I defined `PYTEST_FLAGS` here - it's in the same boat as `PYTEST_TARGET`. I'll take a second look and see if I can't remove it.

> If not, you can use [`passenv`](https://tox.wiki/en/latest/config.html#pass_env) to list envvars that _should_ pass through tox (you could probably move `PYTEST_FLAGS` there as well).

Good to know!

Reply:
> I don't totally remember why I needed to do it this way, but you can also supply those on the command line if you don't want them in `pyproject.toml`:
> 
> https://github.com/uclahs-cds/tool-BoutrosLab-utils/blob/98c270d92007f07de32869770b72ad099aeae5c9/.github/workflows/pytest.yaml#L56-L58
> 
> ```
>           tox \
>             -x 'testenv.pass_env=HOME,GITHUB_TOKEN,GIT_ASKPASS' \
>             -e py${{matrix.python-version}}
> ```

Thanks for the tip. In this case I want to remove as much difference in execution as I can. Whether the commands come from a user running `pytest` or `tox` or `make test`, I want them to all behave the same by default.

Reply:
> In testing, `PYTEST_TARGET` is passed through as expected.

Ahh, I see - I think the difference is that `pytest {env:PYTEST_FLAGS}` works, but trying to access `os.environ["PYTEST_FLAGS"]` within pytest would _not_. `{env:PYTEST_FLAGS}` is expanded _by_ tox so it doesn't need to be passed through.

Reply:
> trying to access `os.environ["PYTEST_FLAGS"]` within pytest would _not_.

Actually, because you're setting `PYTEST_FLAGS` in `setenv`, `os.environ["PYTEST_FLAGS"]` would work. Going back to my original point, `os.environ["PYTEST_TARGET"]` would not work (I'm 99% sure this time).

Reply:
That makes sense. So Tox's behavior is to 

> > trying to access `os.environ["PYTEST_FLAGS"]` within pytest would _not_.
> 
> Actually, because you're setting `PYTEST_FLAGS` in `setenv`, `os.environ["PYTEST_FLAGS"]` would work. Going back to my original point, `os.environ["PYTEST_TARGET"]` would not work (I'm 99% sure this time).

I had to know why this is. Using `strace -s 8192 -o strace.tox.out -ff sh -c 'FOOBARBAZ=abc123 tox'`:

My shell creates `tox` with a copy of my env (note "54 vars").

`execve("/home/aholmes/repos/BL_Python/.venv/bin/tox", ["tox"], 0x55b31f74aed8 /* 54 vars */) = 0`

Tox eventually creates a process with `clone3`, which does not copy the parent process env.

`clone3({flags=CLONE_VM|CLONE_FS|CLONE_FILES|CLONE_SIGHAND|CLONE_THREAD|CLONE_SYSVSEM|CLONE_SETTLS|CLONE_PARENT_SETTID|CLONE_CHILD_CLEARTID, child_tid=0x7fc443fff990, parent_tid=0x7fc443fff990, exit_signal=0, stack=0x7fc4437ff000, stack_size=0x7fff80, tls=0x7fc443fff6c0} => {parent_tid=[839]}, 88) = 839`

This process then runs `pytest` with vfork,

`vfork()                                 = 842`

but the original env that ran Tox (my shell) has been modified by this point, and the envvars in that original env are not passed through (note "16 vars")

`execve("/home/aholmes/repos/BL_Python/.venv/bin/pytest", ["/home/aholmes/repos/BL_Python/.venv/bin/pytest", "--cov=."], 0x7fc4484d5f20 /* 16 vars */) = 0
`

Reply:
![giphy-downsized](https://github.com/user-attachments/assets/cd6b1a49-39b5-46fc-8000-c14e0a6148ed)